ompi/mca/coll/han/coll_han_gatherv.c

```diff
@@ -210,7 +207,9 @@ int mca_coll_han_gatherv_intra(const void *sbuf, int scount, struct ompi_datatyp
         }
 
         if (need_bounce_buf) {
-            bounce_buf = malloc(rdsize * total_up_rcounts);
+            ptrdiff_t rsize, rgap;
```

Comment:
Why do you care about this type ? This process is just an intermediary, for what he cares about he is just manipulating bytes. Of course if you go the MPI_BYTE route you cannot use `ompi_datatype_copy_content_same_ddt` instead you need to pack into a contiguous buffer with the correct size and at the correct index.

Reply:
> It is technically illegal in MPI to send one type and receive it into another type, with the exception of MPI_PACKED.

Understood. I read that in the MPI standard and I think it is indeed the case for `MPI_Gatherv`. But inside ompi's implementation I figured that packing(on sender side) and unpacking(on receiver side) are already taken care of, which means we can indeed break away from the standard when calling `coll_gatherv`.

> this algorithm is not working in a heterogeneous environment
I need your help to understand this statement. What is a heterogeneous environment?

Reply:
> This process is just an intermediary, for what he cares about he is just manipulating bytes. Of course if you go the MPI_BYTE route you cannot use ompi_datatype_copy_content_same_ddt instead you need to pack into a contiguous buffer with the correct size and at the correct index.

In this implementation we only use `MPI_BYTE` on node leaders(excluding the Root) since they don't know the type signature of the other processes - the root and node followers still use the true recv/send datatypes.

So overall the gatherv workflow looks like this:
- Node followers post send with the true send datatype. The data is packed automatically.
- The node leader posts recv with `MPI_BYTE` and stores node followers' packed data on a contiguous temp buffer. Then it posts send with `MPI_BYTE` from the same temp buffer - the data is already packed.
- The Root posts recv with the true recv datatype, and unpacks the data from the node leader into the recv buffer.

In other words, the node leader is transparent and only passes through the packed data from node followers to the Root.

Did I miss anything?

Reply:
It's kind of complicated. OMPI does the data conversion on the receiver, assuming the sender used a compatible datatype (aka. same type signature). So, the sender packs it's local representation into a packed representation, that once received will be unpacked using the sender configuration. And this is where the current scheme will break, because the `MPI_SOURCE` of the receive is not the initial source (where the data has been packed), but the intermediary node. This result in incorrect result after the unpacking.

Let me take an example, 3 nodes with 2 different endianness. A little endian, and B and C big endian. B will be out middleman here. When C receives the data it will come from B, which has the same endianness, so it incorrectly assumes there is nothing to be done. But the data comes in the format of A, which is of a different endianness, so it should have been converted. Long story short, in order to support heterogeneous environments, the middleman need to convert the data between the source and the local format, basically using the local datatype (with the correct count) in both, local and upstream, gatherv. While correct, this approach will be more costly, including for homogeneous environments using non-contiguous datatypes.

Reply:
Thank you very much for the explanation! I wasn't aware of this issue.

> A little endian, and B and C big endian. B will be out middleman here. When C receives the data it will come from B, which has the same endianness, so it incorrectly assumes there is nothing to be done.

I have a couple follow up questions about this.

- What happens when B receives `MPI_BYTE` from A, which has a different endianness? I imagine the receiver, B, will figure out that A is different and "homogenize" the incoming data, so the data will be in B's endianness. In this case when B sends the data to A there shouldn't be additional conversion. I have no clue about this...
- A more detailed question - how does the receiver detect that the data is from a heterogenoeus sender?

Reply:
@bosilca I also realized that this might not be a problem because the node leader and followers are on the same node so they should be equivalent....

Reply:
@bosilca I reckon that this is a problem for scatterv(node leader receives MPI_BYTE from Root which might have a different endianness) but not gatherv(node leader receives MPI_BYTE from node followers which have the same endianness). Therefore I added a homogeneity check in scatterv and bail in the heterogeneous case.

I also considered an alternative, i.e. when node leader receives MPI_BYTE(or similarly MPI_PACKED) from Root, it can swap bytes according to the convertor flags. However, after the discussion in slack I realized this only addresses the endianness conversion, and we also need to ensure the data type's byte representation be correctly translated. With everything considered, the heterogeneous case is rather rare in the field so I decided to be lazy and fall back to the other algorithm.